File: packages/server/remote/src/thumbnails.ts
/**
 * SkyFi MCP Worker — Thumbnail fetcher
 *
 * Fetches satellite image thumbnails in parallel for embedding in MCP
 * responses as base64-encoded ImageContent.
 */

interface ThumbnailRequest {
  archiveId: string;
  url: string;
}

export interface FetchedThumbnail {
  base64: string;
  mimeType: string;
}

/**
 * Fetch thumbnails in parallel with per-image timeouts.
 *
 * @param urls      Array of { archiveId, url } to fetch.
 * @param maxCount  Maximum number of thumbnails to fetch (default 5).
 * @param timeoutMs Per-image timeout in milliseconds (default 5000).
 * @returns Map of archiveId -> FetchedThumbnail. Failed fetches are omitted.
 */
export async function fetchThumbnails(
  urls: ThumbnailRequest[],
  maxCount: number = 5,
  timeoutMs: number = 5000,
): Promise<Map<string, FetchedThumbnail>> {
  if (urls.length === 0) {
    return new Map();
  }

  const toFetch = urls.slice(0, maxCount);

  const results = await Promise.allSettled(
    toFetch.map(async ({ archiveId, url }) => {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), timeoutMs);

      try {
        const resp = await fetch(url, { signal: controller.signal });
        if (!resp.ok) {
          return null;
        }
        const buffer = await resp.arrayBuffer();
        const base64 = arrayBufferToBase64(buffer);
        const mimeType = detectMimeType(new Uint8Array(buffer));
        return { archiveId, base64, mimeType };
      } catch {
        // Timeout, network error, etc. — silently omit
        return null;
      } finally {
        clearTimeout(timer);
      }
    }),
  );

  const thumbnails = new Map<string, FetchedThumbnail>();
  for (const result of results) {
    if (result.status === "fulfilled" && result.value) {
      thumbnails.set(result.value.archiveId, {
        base64: result.value.base64,
        mimeType: result.value.mimeType,
      });
    }
  }
  return thumbnails;
}

/** Convert an ArrayBuffer to a base64 string. */
function arrayBufferToBase64(buffer: ArrayBuffer): string {
  const bytes = new Uint8Array(buffer);
  let binary = "";
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary);
}

/** Detect image MIME type from magic bytes. Defaults to image/png. */
function detectMimeType(bytes: Uint8Array): string {
  if (bytes[0] === 0xff && bytes[1] === 0xd8) return "image/jpeg";
  if (
    bytes[0] === 0x89 && bytes[1] === 0x50 && bytes[2] === 0x4e &&
    bytes[3] === 0x47 && bytes[4] === 0x0d && bytes[5] === 0x0a &&
    bytes[6] === 0x1a && bytes[7] === 0x0a
  ) return "image/png";
  if (
    bytes[0] === 0x52 && bytes[1] === 0x49 && bytes[2] === 0x46 &&
    bytes[3] === 0x46 && bytes[8] === 0x57 && bytes[9] === 0x45 &&
    bytes[10] === 0x42 && bytes[11] === 0x50
  ) return "image/webp";
  if (bytes[0] === 0x47 && bytes[1] === 0x49 && bytes[2] === 0x46) return "image/gif";
  return "image/png";
}

File: packages/core/skyfi_mcp/thumbnails.py
"""
Thumbnail fetching utility.

Fetches satellite image thumbnails in parallel for embedding in MCP responses.
Uses httpx.AsyncClient (already a project dependency) with per-image timeouts.
"""

from __future__ import annotations

import asyncio
import base64
import logging

import httpx

logger = logging.getLogger(__name__)


async def fetch_thumbnails(
    urls: list[tuple[str, str]],
    max_count: int = 5,
    timeout_seconds: float = 5.0,
) -> dict[str, bytes]:
    """Fetch thumbnail images in parallel and return raw bytes.

    Args:
        urls: List of (archive_id, url) pairs to fetch.
        max_count: Maximum number of thumbnails to fetch.
        timeout_seconds: Per-image timeout in seconds.

    Returns:
        Dict mapping archive_id to raw image bytes.
        Failed fetches are silently omitted.
    """
    if not urls:
        return {}

    # Limit to max_count
    to_fetch = urls[:max_count]

    async def _fetch_one(
        client: httpx.AsyncClient, archive_id: str, url: str
    ) -> tuple[str, bytes] | None:
        try:
            resp = await client.get(url, timeout=timeout_seconds)
            resp.raise_for_status()
            return (archive_id, resp.content)
        except (httpx.HTTPError, httpx.TimeoutException):
            logger.debug("Failed to fetch thumbnail for %s: %s", archive_id, url)
            return None

    async with httpx.AsyncClient() as client:
        tasks = [_fetch_one(client, aid, url) for aid, url in to_fetch]
        results = await asyncio.gather(*tasks, return_exceptions=True)

    thumbnails: dict[str, bytes] = {}
    for result in results:
        if isinstance(result, tuple):
            thumbnails[result[0]] = result[1]
    return thumbnails


def encode_thumbnail_base64(data: bytes) -> str:
    """Encode raw image bytes as base64 string."""
    return base64.b64encode(data).decode("ascii")
